nixUnstable: tests fail on darwin due to long $TMPDIR

## Issue description

Building `nixUnstable` on macOS fails in the test phase, since the path to the nix-daemon's socket is longer than the maximum 108 characters.

This is a combination of macOS (long $TMPDIR) and building through nix ($TMPDIR set to the build directory). I'm not really sure whether this is more of a nixpkgs or a nix issue.

The problem can be worked around by setting TMPDIR=/tmp manually, but it seems to me that that should not be required...

Compare nix commit https://github.com/NixOS/nix/commit/c7d44bad000d5f431c05f6c27718d26d2b93bb0f, which removed a workaround for this limitation.

### Steps to reproduce

```
$ uname -a
Darwin aaaa 16.6.0 Darwin Kernel Version 16.6.0: Fri Apr 14 16:21:16 PDT 2017; root:xnu-3789.60.24~6/RELEASE_X86_64 x86_64
$ echo $TMPDIR
/var/folders/rs/k3cmq1b12238c_kx1xg1pndw0000gn/T/
$ nix-build --option use-binary-caches false -A nixUnstable
...
error: socket path ‘/private/var/folders/rs/k3cmq1b12238c_kx1xg1pndw0000gn/T/nix-build-nix-1.12pre5413_b4b1f452.drv-0/nix-test/var/nix/daemon-socket/socket’ is too long
...
FAIL: tests/remote-store.sh
...
error: socket path ‘/private/var/folders/rs/k3cmq1b12238c_kx1xg1pndw0000gn/T/nix-build-nix-1.12pre5413_b4b1f452.drv-0/nix-test/var/nix/daemon-socket/socket’ is too long
...
FAIL: tests/secure-drv-outputs.sh
...
2 out of 42 tests failed 
make: *** [mk/tests.mk:11: installcheck] Error 1
builder for ‘/nix/store/3ixw1yg7h1wnp6ha0rj1b68pp5j25lpp-nix-1.12pre5413_b4b1f452.drv’ failed with exit code 2
error: build of ‘/nix/store/3ixw1yg7h1wnp6ha0rj1b68pp5j25lpp-nix-1.12pre5413_b4b1f452.drv’ failed
```

## Technical details

* System: Darwin 16.6.0
* Nix version: 1.11.11 (building nix-1.12pre5413_b4b1f452)
* Nixpkgs version: master as of 2017-07-04
* Sandboxing enabled: no

--- a/pkgs/tools/package-management/nix/default.nix
+++ b/pkgs/tools/package-management/nix/default.nix
@@ -79,6 +79,9 @@ let
 
     doInstallCheck = true;
 
+    # socket path becomes too long otherwise
+    preInstallCheck = lib.optional stdenv.isDarwin "export TMPDIR=/tmp";
+
     separateDebugInfo = stdenv.isLinux;
 
     crossAttrs = {

